### PR TITLE
cmake: only enable assembly options when SDL_ASSEMBLY is true-ish + copy autoconf loongarch support

### PR DESCRIPTION
**Pull request by madebr**
- Only mark `SDL_xxx` assembly options as configurable by user when `SDL_ASSEMBLY` is true + the arch is correct
- Port the loongarch support to CMake
  @HecaiYuan Could you please test whether the CMake build system can build SDL3 with loongarch lsx/lasx intrinsics?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -438,18 +436,21 @@ set_option(SDL3_DISABLE_UNINSTALL  "Disable uninstallation of SDL3" OFF)
...
+dep_option(SDL_SSE                 "Use SSE assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_SSE2                "Use SSE2 assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_SSE3                "Use SSE3 assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_MMX                 "Use MMX assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_3DNOW               "Use 3Dnow! MMX assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_ALTIVEC             "Use Altivec assembly routines" ON "SDL_ASSEMBLY" OFF)
+dep_option(SDL_ARMSIMD             "Use SIMD assembly blitters on ARM" OFF "SDL_ASSEMBLY;SDL_CPU_ARM32 OR SDL_CPU_ARM64" OFF)
+dep_option(SDL_ARMNEON             "Use NEON assembly blitters on ARM" OFF "SDL_ASSEMBLY;SDL_CPU_ARM32 OR SDL_CPU_ARM64" OFF)
+dep_option(SDL_LSX                 "Use LSX assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_LOONGARCH64" OFF)
+dep_option(SDL_LASX                "Use LASX assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_LOONGARCH64" OFF)
```
Comment (madebr) [inline CMakeLists.txt:450]: @HecaiYuan This assumes lsx/lasx support is only available on loongarch64.
Is this reasonable?

File: cmake/CheckCPUArchitecture.cmake [hunk 2]
```diff
@@ -32,6 +32,8 @@ function(check_cpu_architecture ARCH VARIABLE)
     _internal_check_cpu_architecture("defined(__arm__) || defined(_M_ARM)" arm32 ${VARIABLE})
   elseif(ARCH STREQUAL "arm64")
     _internal_check_cpu_architecture("defined(__aarch64__) || defined(_M_ARM64)" arm64 ${VARIABLE})
+  elseif(ARCH STREQUAL "loongarch64")
+    _internal_check_cpu_architecture("defined(__loongarch64)" loongarch64 ${VARIABLE})
```
Comment (madebr) [inline cmake/CheckCPUArchitecture.cmake:36]: @HecaiYuan Is this a correct way to detect loongarch64?
I got it [from here](https://github.com/gcc-mirror/gcc/blob/0c1b0a23f1fe7db6a2e391b7cb78cff900377772/gcc/config/loongarch/loongarch-c.cc#L71).

File: CMakeLists.txt [hunk 3]
```diff
@@ -438,18 +436,21 @@ set_option(SDL3_DISABLE_UNINSTALL  "Disable uninstallation of SDL3" OFF)
...
+set_option(SDL_ASSEMBLY            "Enable assembly routines" ${OPT_DEF_ASM})
+dep_option(SDL_SSEMATH             "Allow GCC to use SSE floating point math" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_SSE                 "Use SSE assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_SSE2                "Use SSE2 assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_SSE3                "Use SSE3 assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_MMX                 "Use MMX assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_3DNOW               "Use 3Dnow! MMX assembly routines" ON "SDL_ASSEMBLY;SDL_CPU_X86 OR SDL_CPU_X64" OFF)
+dep_option(SDL_ALTIVEC             "Use Altivec assembly routines" ON "SDL_ASSEMBLY" OFF)
+dep_option(SDL_ARMSIMD             "Use SIMD assembly blitters on ARM" OFF "SDL_ASSEMBLY;SDL_CPU_ARM32 OR SDL_CPU_ARM64" OFF)
+dep_option(SDL_ARMNEON             "Use NEON assembly blitters on ARM" OFF "SDL_ASSEMBLY;SDL_CPU_ARM32 OR SDL_CPU_ARM64" OFF)
```
Comment (ccawley2011) [inline CMakeLists.txt:448]:
```suggestion
dep_option(SDL_ARMSIMD             "Use SIMD assembly blitters on ARM" OFF "SDL_ASSEMBLY;SDL_CPU_ARM32" OFF)
dep_option(SDL_ARMNEON             "Use NEON assembly blitters on ARM" OFF "SDL_ASSEMBLY;SDL_CPU_ARM32" OFF)
```

Comment (sezero) [reply CMakeLists.txt:448]: Indeed: those ARM blitters are 32 bits only.

**Comment by slouken** [comment]
Once this is merged, it can be cherry-picked to SDL2.

**Review by sezero — CHANGES_REQUESTED** [review]
OK by me after Cameron's suggestion is made

**Review by sezero — APPROVED** [review]

**Comment by madebr** [comment]
Pushed to SDL2 as edb75bc29 and dcd125236.

**Comment by HecaiYuan** [comment]
- After testing,  HAS_LSX/LASX is not used, it needs to be modified to HAVE_LSX/LASX. 
- Fix loongarch64 support #7031

File: CMakeLists.txt (hunk 1 above)
Comment (HecaiYuan) [reply CMakeLists.txt:452]: No problem.

File: cmake/CheckCPUArchitecture.cmake (hunk 2 above)
Comment (HecaiYuan) [reply cmake/CheckCPUArchitecture.cmake:36]: No problem.

**Pull request merged**